File: src/main/java/io/github/mjaroslav/bon2gradle/task/Bon2Task.java
package io.github.mjaroslav.bon2gradle.task;

import io.github.mjaroslav.bon2gradle.deobf.Bon2ProgressListenerErrorHandler;
import io.github.mjaroslav.bon2gradle.deobf.MappingUtils;
import io.github.mjaroslav.bon2.BON2Impl;
import io.github.mjaroslav.bon2.data.MappingVersion;
import lombok.val;
import org.gradle.api.DefaultTask;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.*;
import org.jetbrains.annotations.UnknownNullability;

import java.io.IOException;

public abstract class Bon2Task extends DefaultTask {
    @InputFile
    public abstract RegularFileProperty getFrom();

    @OutputFile
    public abstract RegularFileProperty getTo();

    @InputFile
    @Optional
    public abstract RegularFileProperty getMapping();

    @Input
    @Optional
    public abstract Property<String> getMappingsRelativeConfPath();

    @TaskAction
    public void doTask() throws IOException {
        val bonHandlers = new Bon2ProgressListenerErrorHandler(getProject());
        val parent = getTo().get().getAsFile().getParentFile();
        if (parent.isDirectory() || parent.mkdirs())
            BON2Impl.remap(getFrom().get().getAsFile(), getTo().get().getAsFile(), getMappingVersion(),
                bonHandlers, bonHandlers);
    }

    @UnknownNullability
    @Internal
    public MappingVersion getMappingVersion() {
        if (!getMapping().isPresent()) return MappingUtils.getCurrentMapping(getProject());
        else return MappingUtils.getMapping(getMapping().get().getAsFile(), getMappingsRelativeConfPath().get());
    }
}

File: src/main/kotlin/task/Bon2Task.kt
package io.github.mjaroslav.bon2gradle.task

import io.github.mjaroslav.bon2.BON2Impl
import io.github.mjaroslav.bon2.data.MappingVersion
import io.github.mjaroslav.bon2gradle.deobf.Bon2ProgressListenerErrorHandler
import io.github.mjaroslav.bon2gradle.deobf.getCurrentMapping
import io.github.mjaroslav.bon2gradle.deobf.getMapping
import org.gradle.api.DefaultTask
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.*

abstract class Bon2Task : DefaultTask() {
    @get:InputFile
    val from: RegularFileProperty = project.objects.fileProperty()

    @get:OutputFile
    val to: RegularFileProperty = project.objects.fileProperty()

    @get:InputFile @Optional
    val mapping: RegularFileProperty = project.objects.fileProperty()

    @get:Input @Optional
    val mappingRelativeConfPath: Property<String> = project.objects.property(String::class.java)

    @Internal
    fun getMappingVersion(): MappingVersion {
        return if (!mapping.isPresent) getCurrentMapping(project)
        else getMapping(mapping.get().asFile, mappingRelativeConfPath.get())
    }

    @TaskAction
    fun doTask() {
        val bonHandlers = Bon2ProgressListenerErrorHandler(project.logger)
        val parent = to.get().asFile.parentFile
        if (parent.isDirectory || parent.mkdirs())
            BON2Impl.remap(from.get().asFile, to.get().asFile, getMappingVersion(), bonHandlers, bonHandlers)
    }
}
